### PR TITLE
Validate the contents of the VTX Json when loading

### PR DESCRIPTION
**Pull request by McGiverGim**
Some users seems to have problems while loading a JSON in the VTX Table tab.

The code validates that is comes in a correct JSON format, but does not validate the contents.

This PR is to discuss what is the best approach.
1. This PR uses a simple code to detect the values of each field. If the field is not the correct type it throws an error. This comes another question: I can replace the bad value with other correct empty value (empty string, zero,etc.) What is the best solution?
2. Another approach, more elegant but I don't know if better, is to define a JSON Schema and try to validate against it.

When the JSON is wrong or the contents are wrong, a message appears in the upper log. Other solution can be show an error message more clear in the tab,but now there are other messages in the tab too and maybe they are too much.

### CODE REVIEW

File: src/js/tabs/vtx.js [hunk 1]
```diff
@@ -101,19 +101,46 @@ TABS.vtx.initialize = function (callback) {
     // Emulates the MSP read from a vtxConfig object (JSON)
     function read_vtx_config_json(vtxConfig, vtxcallback_after_read) {
 
+        function checkArrayType(value, type) {
+
+            // Check each of its elements
+            if (Array.isArray(value)) {
+                value.forEach(function(element) {
```
Comment (Anisotropic) [inline src/js/tabs/vtx.js]: value.forEach((el) => checkType(el, type));
is more readable

Comment (McGiverGim) [reply src/js/tabs/vtx.js]: True, now I'm out of the computer, I will change it later

Comment (Anisotropic) [reply src/js/tabs/vtx.js]: It's not necessary. Just my 2 cent)

Comment (McGiverGim) [reply src/js/tabs/vtx.js]: Fixed!

File: src/js/tabs/vtx.js [hunk 2]
```diff
@@ -101,19 +101,44 @@ TABS.vtx.initialize = function (callback) {
     // Emulates the MSP read from a vtxConfig object (JSON)
     function read_vtx_config_json(vtxConfig, vtxcallback_after_read) {
 
+        function checkArrayType(value, type) {
+
+            // Check each of its elements
+            if (Array.isArray(value)) {
+                value.forEach((element) => checkType(element, type));
+            } else {
+                let error = "Error in array sintax of the JSON";
```
Comment (mikeller) [inline src/js/tabs/vtx.js]: 'syntax'

**Comment by mikeller** [comment]
I think this will be helpful in catching some / most of the potential problems that the current implementation has, but it is not a particularly nice or portable.

I think for longer term we should be looking if there it is possible to implement this based on JSON schema validation with reasonable effort.

**Comment by McGiverGim** [comment]
I don't know if I will have time the next days, but the only difficult part is to look for a JavaScript library that does the validation. There are several of them.

**Comment by mikeller** [comment]
@McGiverGim: I did some digging for you, and looks like we've got a good candidate here: https://www.npmjs.com/package/ajv#using-in-browser

**Comment by McGiverGim** [comment]
True! I will try to look into it in the next week's.

**Comment by McGiverGim** [comment]
@mikeller added validation against JSON Schema. I'm not too sure if the folder to store the schemas is correct, I've created it in: ` src/resources/jsonschema/`

**Comment by McGiverGim** [comment]
Rebased against master to test against Sonar 😋

**Comment by McGiverGim** [comment]
It seems the new `ajv` library has a lot of issues in Sonar. My code seems clean.

**Comment by mikeller** [comment]
@McGiverGim: Adding a copy of the library into the repository will make keeping it up to date harder than it needs be.
It is easier and preferrable to add it with 

```
yarn add ajv
```

And then include the minified version with

```
<script type="text/javascript" src="./node_modules/ajv/dist/ajv.min.js"></script>
```

Also, good work on the schema validation!

**Comment by McGiverGim** [comment]
I can't. It does not work for the chrome app. This is an special version for browsers.

**Comment by mikeller** [comment]
@McGiverGim: Heh, it uses `eval()`. While this isnt't good practice, and should normally be avoided, it is commonplace for legacy JavaScript to use it to jury-rig a version of templating.

That's annoying, and I do not know what to best do here - adding more libraries into the repository is not a good approach, especially for ones that have security implications, as upgrading them will be hard / prone to introduce errors.

Maybe we should just decide to pull the plug on Chrome apps. The usage numbers are only around 0.3% of all users:

image

**Comment by McGiverGim** [comment]
True. I don't know if we can find another library compatible to validate the JSON. I didn't find one but I only searched some minutes.

**Comment by Docteh** [comment]
If we discontinue the chrome os build can we keep the Jenkins build? Or maybe explicitly package as `dist`? I had a jokey idea where if a user had a Jenkins build running as a portable app, it could download latest version while in the middle of starting up.

**Comment by McGiverGim** [comment]
If it works I suppose we can maintain a dist build.

**Comment by mikeller** [comment]
@Docteh: Why would we want to keep the Jenkins build if the app that is built does not run any more? This will just be an endless source of user frustration and support issues.

We should potentially put more effort into getting Windows / MacOS build pipelines going,like what we are working on in #1727. 

Also, I am not sure that we want to drive up the usage of nightly builds of configurator - we are already getting 'support requests' from users who are using nightly versions but don't read up on the issue trackers, and then open duplicate issues for stuff that is still under development. Unless you are a developer, or you want to beta test a new feature and are willing to spend the time to read up on the issue tracker, you should not be using unreleased versions.

**Comment by Docteh** [comment]
@mikeller Windows builds will make me feel better about it, but right now grabbing a copy of the chromeos zip, and taping it together with NW.js is a nice easy way to get a windows build. Want a newer build? download a newer build and dump it over the older build.

**Comment by mikeller** [comment]
@Docteh: This may be an acceptable approach for a developer, but I would not want to ever promote this as a 'supported' way of running configurator - too many things that can go wrong, and too much potential for errors to pop up that will leave the non-developer user up the creek without a paddle.

And as a developer, just clone the repository, and then do `git pull; yarn start` - done. :wink:

**Comment by Docteh** [comment]
https://github.com/epoberezkin/ajv-pack sounds like it'd get around actually calling eval.

**Comment by mikeller** [comment]
@Docteh: Yes, pre-evaluating the JSON schema as part of the build pipeline would get us around the problem of not being able to do this in the Chrome app. But that would require extra steps to be added / maintained in the build pipeline, which is not attractive. Let's first explore the 'ditch ChromeOS' approach.

**Comment by McGiverGim** [comment]
@mikeller I have a solution for this (we can remove the ChromeOS too if we want), but I have found a replacement for the `ajv` library, that can be imported through npm and works in the browser too: `djv` (https://github.com/korzio/djv).

It seems to be faster and more precise too:
https://github.com/ebdrup/json-schema-benchmark

I have pushed the changes to this branch. But it uses `eval` too so if we want to make it work against Chrome I think we need to modify the `manifest.json` too.

**Comment by McGiverGim** [comment]
@mikeller, sorry, but maybe I did not explain it well... this can work with ChromeOS, but we need to modify the `manifest.json` to add the `eval-unsafe` right. I'm not too sure how to add it exactly, so my code did not have it, so at this moment we have broken the ChromeOS VTX load from file :(

I can remove the validation for Chrome OS, or try to find the correct way to add the right in the manifest.

**Comment by mikeller** [comment]
@McGiverGim: djv uses eval as well? Sorry, I did not even check. :disappointed: 

>  this can work with ChromeOS, but we need to modify the manifest.json to add the `eval-unsafe` right. I'm not too sure how to add it exactly,

For all that I could find this is not possible for chrome apps - it is possible for extensions by setting `content_security_policy`, but this setting is not supported for Chrome apps - most likely intentionally by Google, as they want to phase out Chrome apps.

**Comment by McGiverGim** [comment]
I can remove the JSON Schema validation for Chrome OS at this moment, or if someone can find another JSON Schema validator I can play with it...

**Comment by mikeller** [comment]
@McGiverGim: That sounds like a reasonable approach - we really need to make a decision re discontinuing support for the Chrome app builds.

**Comment by McGiverGim** [comment]
@mikeller done: https://github.com/betaflight/betaflight-configurator/pull/1802

**Pull request merged**